### DISCUSSION TITLE
Read or write over files

### DISCUSSION THREAD

**Discussion by tisonkun** [netty/netty#13716] [Q&A] [answered]
This is a newcomer's question for the usage of Netty.

AFAIK, Netty is built upon Sockets and thus can be used for network IO. While sockets and regular files are both represented as FD from the OS perspective, I wonder if Netty can perform read or write over files.

Epoll / IOCP / io_uring can be used in both network IO and file IO.

In another words, compare the Netty's architecture and Asio's, I'm curious about the reason for the differences esp. for sockets vs. generic FDs:

![image](https://github.com/netty/netty/assets/18818196/7d27b051-71ad-47b5-a82d-eaf7eda86dd4)

![image](https://github.com/netty/netty/assets/18818196/9cb18536-4b72-4e4f-9f00-bc0ee3c82795)

Anyway, it seems Netty is an attempt to improve JDK's sockets IO but I don't find widely-used libraries that tries to improve JDK's files IO.

**Comment by chrisvest** [comment, marked as answer]
Netty's programming model is centered around streaming data or messages in and out, which mirrors the stream and datagram sockets APIs.
Files, on the other hand, are seekable, extendable byte arrays. This favors a different programming model, and it's awkward to fit such APIs into the Netty pipeline/handler model.

**Reply by tisonkun** [reply]
Thank you! Out of the thread, I wonder if you're aware any FS libs for Java, perhaps leverage low-level FS API like Netty is built upon native interfaces.

**Reply by chrisvest** [reply]
No, I don't know of any libraries like that. I think everyone just uses the built in Java ones.
